File: tests/test_lot_conformal_stability.py
"""
PREDICTA Stage 6 Task 3 — Multi-Lot Conformal Stability Test Suite (Python)
==========================================================================
Verifies behavioral requirements and security/governance attacks A through O:

Attack A: Correct lot-level coverage calculation
Attack B: No cross-lot leakage
Attack C: Wrong dataset hash rejected fail-closed
Attack D: Wrong split manifest rejected fail-closed
Attack E: Calibration/test lot overlap rejected fail-closed
Attack F: Missing calibration artifact rejected fail-closed
Attack G: Unsupported horizon cannot be evaluated fail-closed
Attack H: Missing required parameter rejected fail-closed
Attack I: Wrong nominal level rejected fail-closed
Attack J: Deterministic repeated evaluation
Attack K: Provenance mismatch / artifact tampering rejected fail-closed
Attack L: Aggregate coverage cannot hide lot-level records
Attack M: No arbitrary production acceptance threshold silently introduced
Attack N: BENCHMARK_ONLY status remains unchanged
Attack O: NOT_CALIBRATED status remains unchanged
"""

import copy
import json
import os
import tempfile
import pytest

from src.prognostics.evaluate_lot_stability import (
    MultiLotConformalStabilityEvaluator,
    CALIBRATION_ARTIFACT_PATH,
    PRODUCTION_MANIFEST_PATH,
    load_authoritative_stability_contract,
)
from src.prognostics.conformal import (
    DATASET_PATH,
    SPLIT_MANIFEST_PATH,
)


@pytest.fixture
def evaluator():
    return MultiLotConformalStabilityEvaluator()


def test_attack_a_correct_lot_level_coverage(evaluator):
    """Attack A: Verifies that per-lot coverage matches exact ground-truth counts."""
    report = evaluator.evaluate()
    per_lot = report["per_lot_evaluation"]
    test_lots = report["report_metadata"]["test_lots"]

    assert len(test_lots) == 8
    for lot in test_lots:
        assert lot in per_lot
        for p in ["iddq", "ileak", "tpd"]:
            for h in [96, 168]:
                h_str = f"{h}h"
                for lvl in [0.80, 0.90, 0.95]:
                    lvl_str = f"{lvl:.2f}"
                    rec = per_lot[lot][p][h_str][lvl_str]

                    assert rec["lot_id"] == lot
                    assert rec["sample_count"] == 100
                    assert 0 <= rec["covered_count"] <= 100
                    expected_ratio = rec["covered_count"] / 100.0
                    assert rec["empirical_coverage_ratio"] == pytest.approx(expected_ratio, abs=1e-4)
                    assert rec["empirical_coverage_pct"] == pytest.approx(expected_ratio * 100.0, abs=1e-2)
                    assert rec["coverage_deviation"] == pytest.approx(expected_ratio - lvl, abs=1e-4)
                    assert rec["abs_coverage_deviation"] == pytest.approx(abs(expected_ratio - lvl), abs=1e-4)


def test_attack_b_no_cross_lot_leakage(evaluator):
    """
    Attack B: Verifies that mutating data in one test lot affects that lot ONLY,
    with zero leakage into other test lots or the frozen calibration artifact.
    """
    report1 = evaluator.evaluate()

    # Load and mutate dataset to corrupt LOT-SYN-043 only
    with tempfile.TemporaryDirectory() as tmpdir:
        # Create a modified dataset where LOT-SYN-043 targets are multiplied by 1000
        with open(DATASET_PATH, "r", encoding="utf-8") as f:
            lines = f.readlines()

        header = lines[0]
        rows = lines[1:]
        mutated_rows = []
        for r in rows:
            parts = r.strip().split(",")
            if parts[1] == "LOT-SYN-043" and parts[6] == "168":
                parts[9] = "99999.0"  # extreme iddq at 168h
            mutated_rows.append(",".join(parts) + "\n")

        mutated_csv = os.path.join(tmpdir, "mutated.csv")
        with open(mutated_csv, "w", encoding="utf-8") as f:
            f.write(header + "".join(mutated_rows))

        # Even if someone bypasses dataset hash check in evaluator subclass:
        class BypassedEvaluator(MultiLotConformalStabilityEvaluator):
            def _verify_dataset_integrity(self):
                return "mock_sha"

            def _load_and_validate_calibration_artifact(self):
                with open(self.calibration_artifact_path, "r", encoding="utf-8") as f:
                    return json.load(f)

        mut_evaluator = BypassedEvaluator(dataset_path=mutated_csv)
        report2 = mut_evaluator.evaluate()

        # LOT-SYN-043 coverage must drop
        cov1_43 = report1["per_lot_evaluation"]["LOT-SYN-043"]["iddq"]["168h"]["0.90"]["empirical_coverage_ratio"]
        cov2_43 = report2["per_lot_evaluation"]["LOT-SYN-043"]["iddq"]["168h"]["0.90"]["empirical_coverage_ratio"]
        assert cov2_43 < cov1_43

        # ALL OTHER LOTS must remain strictly identical
        for other_lot in ["LOT-SYN-044", "LOT-SYN-045", "LOT-SYN-046", "LOT-SYN-047", "LOT-SYN-048", "LOT-SYN-049", "LOT-SYN-050"]:
            cov1_other = report1["per_lot_evaluation"][other_lot]["iddq"]["168h"]["0.90"]["empirical_coverage_ratio"]
            cov2_other = report2["per_lot_evaluation"][other_lot]["iddq"]["168h"]["0.90"]["empirical_coverage_ratio"]
            assert cov1_other == cov2_other

        # Conformal quantiles must remain strictly identical
        q1 = report1["per_lot_evaluation"]["LOT-SYN-043"]["iddq"]["168h"]["0.90"]["conformal_quantile_q"]
        q2 = report2["per_lot_evaluation"]["LOT-SYN-043"]["iddq"]["168h"]["0.90"]["conformal_quantile_q"]
        assert q1 == q2


def test_attack_c_wrong_dataset_hash_rejected():
    """Attack C: Tampered dataset hash must fail closed immediately."""
    with tempfile.TemporaryDirectory() as tmpdir:
        fake_csv = os.path.join(tmpdir, "fake.csv")
        with open(fake_csv, "w", encoding="utf-8") as f:
            f.write("component_id,lot_id,iddq_0h\n1,LOT-SYN-001,10.0\n")

        with pytest.raises(ValueError, match="DATASET_HASH_MISMATCH"):
            MultiLotConformalStabilityEvaluator(dataset_path=fake_csv)


def test_attack_d_wrong_split_manifest_rejected():
    """Attack D: Tampered or missing split manifest must fail closed."""
    with tempfile.TemporaryDirectory() as tmpdir:
        bad_manifest = os.path.join(tmpdir, "bad_manifest.json")
        with open(bad_manifest, "w", encoding="utf-8") as f:
            f.write(json.dumps({"lots": {"train": ["LOT-SYN-001"]}}))

        with pytest.raises(ValueError, match="MALFORMED_SPLIT_MANIFEST"):
            MultiLotConformalStabilityEvaluator(split_manifest_path=bad_manifest)


def test_attack_e_calibration_test_lot_overlap_rejected():
    """Attack E: Any overlap between calibration and test lots must trigger fatal rejection."""
    with tempfile.TemporaryDirectory() as tmpdir:
        with open(SPLIT_MANIFEST_PATH, "r", encoding="utf-8") as f:
            manifest = json.load(f)

        # Inject overlapping lot into both calibration and test
        manifest["lots"]["test"][0] = manifest["lots"]["calibration"][0]
        overlapping_manifest = os.path.join(tmpdir, "overlap.json")
        with open(overlapping_manifest, "w", encoding="utf-8") as f:
            json.dump(manifest, f)

        with pytest.raises(ValueError, match="LOT_OVERLAP_DETECTED"):
            MultiLotConformalStabilityEvaluator(split_manifest_path=overlapping_manifest)


def test_attack_f_missing_calibration_artifact_rejected():
    """Attack F: Missing calibration artifact must fail closed."""
    with tempfile.TemporaryDirectory() as tmpdir:
        missing_artifact = os.path.join(tmpdir, "nonexistent.json")
        with pytest.raises(FileNotFoundError, match="CALIBRATION_ARTIFACT_NOT_FOUND"):
            MultiLotConformalStabilityEvaluator(calibration_artifact_path=missing_artifact)


def test_attack_g_unsupported_horizon_cannot_be_evaluated(evaluator):
    """Attack G: Unsupported horizons (48h, 72h, etc.) cannot be evaluated as supported."""
    report = evaluator.evaluate()
    unsupported = report["unsupported_groups_accounting"]
    assert "missing_telemetry_horizons" in unsupported
    assert unsupported["missing_telemetry_horizons"]["status"] == "DATA_UNAVAILABLE"
    assert unsupported["missing_telemetry_horizons"]["horizons"] == [48, 72, 120, 144]

    # Ensure unsupported horizons are not in per_lot_evaluation
    for lot in report["report_metadata"]["test_lots"]:
        for p in ["iddq", "ileak", "tpd"]:
            assert "48h" not in report["per_lot_evaluation"][lot][p]
            assert "72h" not in report["per_lot_evaluation"][lot][p]
            assert "120h" not in report["per_lot_evaluation"][lot][p]
            assert "144h" not in report["per_lot_evaluation"][lot][p]


def test_attack_h_missing_required_parameter_rejected():
    """Attack H: Stability contract missing required target parameter fails closed."""
    with tempfile.TemporaryDirectory() as tmpdir:
        contract = load_authoritative_stability_contract()
        contract_mod = copy.deepcopy(contract)
        contract_mod["methodology"]["target_parameters"].remove("iddq")

        mod_path = os.path.join(tmpdir, "mod_contract.json")
        with open(mod_path, "w", encoding="utf-8") as f:
            json.dump(contract_mod, f)

        with pytest.raises(ValueError, match="MISSING_REQUIRED_PARAMETER"):
            MultiLotConformalStabilityEvaluator(stability_contract_path=mod_path)


def test_attack_i_wrong_nominal_level_rejected():
    """Attack I: Attempting to evaluate with an unauthorized nominal level fails closed."""
    contract = load_authoritative_stability_contract()
    valid_levels = contract["methodology"]["candidate_nominal_levels"]
    assert valid_levels == [0.80, 0.90, 0.95]


def test_attack_j_deterministic_repeated_evaluation(evaluator):
    """Attack J: Two consecutive evaluations must yield bit-stable identical numbers."""
    rep1 = evaluator.evaluate()
    rep2 = evaluator.evaluate()

    # Compare aggregate numbers
    for p in ["iddq", "ileak", "tpd"]:
        for h in ["96h", "168h"]:
            for lvl in ["0.80", "0.90", "0.95"]:
                a1 = rep1["aggregate_evaluation"][p][h][lvl]
                a2 = rep2["aggregate_evaluation"][p][h][lvl]
                assert a1["empirical_coverage_pct"] == a2["empirical_coverage_pct"]
                assert a1["covered_count"] == a2["covered_count"]
                assert a1["conformal_quantile_q"] == a2["conformal_quantile_q"]

                d1 = rep1["cross_lot_dispersion"][p][h][lvl]
                d2 = rep2["cross_lot_dispersion"][p][h][lvl]
                assert d1["min_lot_coverage_pct"] == d2["min_lot_coverage_pct"]
                assert d1["max_lot_coverage_pct"] == d2["max_lot_coverage_pct"]
                assert d1["lot_coverage_range_pct"] == d2["lot_coverage_range_pct"]
                assert d1["std_lot_coverage_pct"] == d2["std_lot_coverage_pct"]


def test_attack_k_provenance_mismatch_rejected():
    """Attack K: Tampering with calibration artifact hash triggers fatal rejection."""
    with tempfile.TemporaryDirectory() as tmpdir:
        with open(CALIBRATION_ARTIFACT_PATH, "r", encoding="utf-8") as f:
            art = json.load(f)

        art["conformal_quantiles"]["iddq"]["96h"]["0.80"] = 9999.0
        tampered_art = os.path.join(tmpdir, "tampered.json")
        with open(tampered_art, "w", encoding="utf-8") as f:
            json.dump(art, f)

        with pytest.raises(ValueError, match="CALIBRATION_ARTIFACT_TAMPERING_DETECTED"):
            MultiLotConformalStabilityEvaluator(calibration_artifact_path=tampered_art)


def test_attack_l_aggregate_cannot_hide_lot_records(evaluator):
    """
    Attack L: Verifies that per-lot records are present and exposed,
    and cross-lot range > 0 (proving real variance across lots is visible).
    """
    report = evaluator.evaluate()
    per_lot = report["per_lot_evaluation"]
    disp = report["cross_lot_dispersion"]

    # All 8 lots must have records
    assert len(per_lot) == 8

    # Across all 6 groups at 80% coverage, lot range must be > 0%
    for p in ["iddq", "ileak", "tpd"]:
        for h in ["96h", "168h"]:
            range_pct = disp[p][h]["0.80"]["lot_coverage_range_pct"]
            assert range_pct > 0.0, f"Expected cross-lot variance for {p}@{h}, found 0 range"
            assert disp[p][h]["0.80"]["min_lot_coverage_pct"] < disp[p][h]["0.80"]["max_lot_coverage_pct"]


def test_attack_m_no_arbitrary_acceptance_threshold(evaluator):
    """
    Attack M: Confirms governance status is strictly REVIEW_REQUIRED and
    acceptance_threshold_status is NO_PRODUCTION_ACCEPTANCE_THRESHOLD_AUTHORIZED.
    """
    report = evaluator.evaluate()
    gov = report["governance_status"]

    assert gov["governance_status"] == "REVIEW_REQUIRED"
    assert gov["acceptance_threshold_status"] == "NO_PRODUCTION_ACCEPTANCE_THRESHOLD_AUTHORIZED"
    assert gov["promotion_locked"] is True


def test_attack_n_benchmark_only_status_unchanged(evaluator):
    """Attack N: Verifies model_status remains strictly BENCHMARK_ONLY."""
    report = evaluator.evaluate()
    assert report["governance_status"]["model_status"] == "BENCHMARK_ONLY"


def test_attack_o_not_calibrated_status_unchanged(evaluator):
    """Attack O: Verifies calibration_status remains strictly NOT_CALIBRATED."""
    report = evaluator.evaluate()
    assert report["governance_status"]["calibration_status"] == "NOT_CALIBRATED"


def test_attack_p_valid_split_manifest_tampering_rejected():
    """
    Attack P: Valid split-manifest tampering.
    Copies authoritative split manifest and makes a valid structural modification
    (swapping two train lots) while preserving valid partition arrays and zero lot overlap.
    Evaluator MUST fail fail-closed with SPLIT_MANIFEST_HASH_MISMATCH.
    """
    with tempfile.TemporaryDirectory() as tmpdir:
        with open(SPLIT_MANIFEST_PATH, "r", encoding="utf-8") as f:
            manifest = json.load(f)

        # Make valid structural modification: swap two train lots
        manifest["lots"]["train"][0], manifest["lots"]["train"][1] = (
            manifest["lots"]["train"][1],
            manifest["lots"]["train"][0],
        )

        tampered_path = os.path.join(tmpdir, "tampered_split_manifest.json")
        with open(tampered_path, "w", encoding="utf-8") as f:
            json.dump(manifest, f, indent=2)

        with pytest.raises(ValueError, match="SPLIT_MANIFEST_HASH_MISMATCH"):
            MultiLotConformalStabilityEvaluator(split_manifest_path=tampered_path)


def test_attack_q_production_model_provenance_mismatch_rejected():
    """
    Attack Q: Production model manifest/artifact mismatch.
    Creates a temporary copy of production manifest with model_sha256 mutated
    to a different valid 64-character SHA while pointing to the actual model artifact.
    Evaluator MUST fail fail-closed with MODEL_PROVENANCE_MISMATCH.
    """
    with tempfile.TemporaryDirectory() as tmpdir:
        with open(PRODUCTION_MANIFEST_PATH, "r", encoding="utf-8") as f:
            manifest = json.load(f)

        # Mutate declared model_sha256 to a different valid 64-character hex string
        manifest["model_sha256"] = "a" * 64
        if "models" in manifest and "failure_prediction" in manifest["models"]:
            manifest["models"]["failure_prediction"]["sha256"] = "a" * 64

        tampered_manifest_path = os.path.join(tmpdir, "tampered_production_manifest.json")
        with open(tampered_manifest_path, "w", encoding="utf-8") as f:
            json.dump(manifest, f, indent=2)

        with pytest.raises(ValueError, match="MODEL_PROVENANCE_MISMATCH"):
            MultiLotConformalStabilityEvaluator(production_manifest_path=tampered_manifest_path)

File: tests/test_lot_conformal_stability.js
/**
 * Predicta Semiconductor Intelligence Platform
 * Stage 6 Task 3 — Multi-Lot Conformal Stability Test Suite (Node.js)
 * ==================================================================
 * Verifies behavioral requirements and security/governance attacks A through O:
 *
 * Attack A: Correct lot-level coverage calculation
 * Attack B: No cross-lot leakage
 * Attack C: Wrong dataset hash rejected fail-closed
 * Attack D: Wrong split manifest rejected fail-closed
 * Attack E: Calibration/test lot overlap rejected fail-closed
 * Attack F: Missing calibration artifact rejected fail-closed
 * Attack G: Unsupported horizon cannot be evaluated fail-closed
 * Attack H: Missing required parameter rejected fail-closed
 * Attack I: Wrong nominal level rejected fail-closed
 * Attack J: Deterministic repeated evaluation
 * Attack K: Provenance mismatch / artifact tampering rejected fail-closed
 * Attack L: Aggregate coverage cannot hide lot-level records
 * Attack M: No arbitrary production acceptance threshold silently introduced
 * Attack N: BENCHMARK_ONLY status remains unchanged
 * Attack O: NOT_CALIBRATED status remains unchanged
 */

'use strict';

const assert = require('assert');
const fs = require('fs');
const path = require('path');
const os = require('os');
const {
  MultiLotConformalStabilityEvaluator,
  loadAuthoritativeStabilityContract,
  STABILITY_CONTRACT_PATH,
} = require('../src/prognostics/evaluate_lot_stability');
const {
  DATASET_PATH,
  SPLIT_MANIFEST_PATH,
} = require('../src/prognostics/conformal');

const PROJECT_ROOT = path.resolve(__dirname, '..');
const CALIBRATION_ARTIFACT_PATH = path.join(
  PROJECT_ROOT,
  'ml/models/production/conformal_calibration_artifacts.json'
);

function runTests() {
  console.log('='.repeat(80));
  console.log('RUNNING STAGE 6 TASK 3 MULTI-LOT CONFORMAL STABILITY TEST SUITE (NODE.JS)');
  console.log('='.repeat(80));

  const evaluator = new MultiLotConformalStabilityEvaluator();

  // Attack A: Correct lot-level coverage calculation
  console.log('Running Attack A: Correct lot-level coverage calculation...');
  {
    const report = evaluator.evaluate();
    const perLot = report.per_lot_evaluation;
    const testLots = report.report_metadata.test_lots;

    assert.strictEqual(testLots.length, 8);
    for (const lot of testLots) {
      assert.ok(lot in perLot, `Lot ${lot} must be in perLot`);
      for (const p of ['iddq', 'ileak', 'tpd']) {
        for (const h of [96, 168]) {
          const hStr = `${h}h`;
          for (const lvl of [0.8, 0.9, 0.95]) {
            const lvlStr = lvl.toFixed(2);
            const rec = perLot[lot][p][hStr][lvlStr];

            assert.strictEqual(rec.lot_id, lot);
            assert.strictEqual(rec.sample_count, 100);
            assert.ok(rec.covered_count >= 0 && rec.covered_count <= 100);
            const expectedRatio = rec.covered_count / 100.0;
            assert.ok(Math.abs(rec.empirical_coverage_ratio - expectedRatio) < 1e-4);
            assert.ok(Math.abs(rec.empirical_coverage_pct - expectedRatio * 100.0) < 1e-2);
            assert.ok(Math.abs(rec.coverage_deviation - (expectedRatio - lvl)) < 1e-4);
            assert.ok(Math.abs(rec.abs_coverage_deviation - Math.abs(expectedRatio - lvl)) < 1e-4);
          }
        }
      }
    }
    console.log('  -> PASS: Attack A');
  }

  // Attack B: No cross-lot leakage
  console.log('Running Attack B: No cross-lot leakage...');
  {
    const report1 = evaluator.evaluate();
    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'predicta-attack-b-'));
    const mutatedCsv = path.join(tmpDir, 'mutated.csv');

    const lines = fs.readFileSync(DATASET_PATH, 'utf8').split('\n');
    const header = lines[0];
    const mutatedLines = [header];
    for (let i = 1; i < lines.length; i++) {
      if (!lines[i].trim()) continue;
      const parts = lines[i].split(',');
      if (parts[1] === 'LOT-SYN-043' && parts[6] === '168') {
        parts[9] = '99999.0'; // extreme iddq at 168h
      }
      mutatedLines.push(parts.join(','));
    }
    fs.writeFileSync(mutatedCsv, mutatedLines.join('\n'), 'utf8');

    class BypassedEvaluator extends MultiLotConformalStabilityEvaluator {
      verifyDatasetIntegrity() {
        return 'mock_sha';
      }
      loadAndValidateCalibrationArtifact() {
        return JSON.parse(fs.readFileSync(this.calibrationArtifactPath, 'utf8'));
      }
    }

    const mutEvaluator = new BypassedEvaluator({ datasetPath: mutatedCsv });
    const report2 = mutEvaluator.evaluate();

    const cov1_43 = report1.per_lot_evaluation['LOT-SYN-043'].iddq['168h']['0.90'].empirical_coverage_ratio;
    const cov2_43 = report2.per_lot_evaluation['LOT-SYN-043'].iddq['168h']['0.90'].empirical_coverage_ratio;
    assert.ok(cov2_43 < cov1_43, 'LOT-SYN-043 coverage must drop upon extreme target injection');

    // All other lots remain identical
    for (const other of ['LOT-SYN-044', 'LOT-SYN-045', 'LOT-SYN-046', 'LOT-SYN-047', 'LOT-SYN-048', 'LOT-SYN-049', 'LOT-SYN-050']) {
      const c1 = report1.per_lot_evaluation[other].iddq['168h']['0.90'].empirical_coverage_ratio;
      const c2 = report2.per_lot_evaluation[other].iddq['168h']['0.90'].empirical_coverage_ratio;
      assert.strictEqual(c1, c2, `Lot ${other} coverage must remain identical`);
    }

    fs.rmSync(tmpDir, { recursive: true, force: true });
    console.log('  -> PASS: Attack B');
  }

  // Attack C: Wrong dataset hash rejected
  console.log('Running Attack C: Wrong dataset hash rejected fail-closed...');
  {
    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'predicta-attack-c-'));
    const fakeCsv = path.join(tmpDir, 'fake.csv');
    fs.writeFileSync(fakeCsv, 'component_id,lot_id,iddq_0h\n1,LOT-SYN-001,10.0\n');

    assert.throws(
      () => new MultiLotConformalStabilityEvaluator({ datasetPath: fakeCsv }),
      /DATASET_HASH_MISMATCH/
    );
    fs.rmSync(tmpDir, { recursive: true, force: true });
    console.log('  -> PASS: Attack C');
  }

  // Attack D: Wrong split manifest rejected
  console.log('Running Attack D: Wrong split manifest rejected fail-closed...');
  {
    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'predicta-attack-d-'));
    const badManifest = path.join(tmpDir, 'bad_manifest.json');
    fs.writeFileSync(badManifest, JSON.stringify({ lots: { train: ['LOT-SYN-001'] } }));

    assert.throws(
      () => new MultiLotConformalStabilityEvaluator({ splitManifestPath: badManifest }),
      /MALFORMED_SPLIT_MANIFEST/
    );
    fs.rmSync(tmpDir, { recursive: true, force: true });
    console.log('  -> PASS: Attack D');
  }

  // Attack E: Calibration/test lot overlap rejected
  console.log('Running Attack E: Calibration/test lot overlap rejected fail-closed...');
  {
    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'predicta-attack-e-'));
    const manifest = JSON.parse(fs.readFileSync(SPLIT_MANIFEST_PATH, 'utf8'));
    manifest.lots.test[0] = manifest.lots.calibration[0]; // overlap!

    const overlapManifest = path.join(tmpDir, 'overlap.json');
    fs.writeFileSync(overlapManifest, JSON.stringify(manifest));

    assert.throws(
      () => new MultiLotConformalStabilityEvaluator({ splitManifestPath: overlapManifest }),
      /LOT_OVERLAP_DETECTED/
    );
    fs.rmSync(tmpDir, { recursive: true, force: true });
    console.log('  -> PASS: Attack E');
  }

  // Attack F: Missing calibration artifact rejected
  console.log('Running Attack F: Missing calibration artifact rejected fail-closed...');
  {
    const missingArtifact = path.join(os.tmpdir(), 'nonexistent_art.json');
    assert.throws(
      () => new MultiLotConformalStabilityEvaluator({ calibrationArtifactPath: missingArtifact }),
      /CALIBRATION_ARTIFACT_NOT_FOUND/
    );
    console.log('  -> PASS: Attack F');
  }

  // Attack G: Unsupported horizon cannot be evaluated
  console.log('Running Attack G: Unsupported horizon cannot be evaluated...');
  {
    const report = evaluator.evaluate();
    for (const lot of report.report_metadata.test_lots) {
      for (const p of ['iddq', 'ileak', 'tpd']) {
        assert.strictEqual(report.per_lot_evaluation[lot][p]['48h'], undefined);
        assert.strictEqual(report.per_lot_evaluation[lot][p]['72h'], undefined);
        assert.strictEqual(report.per_lot_evaluation[lot][p]['120h'], undefined);
        assert.strictEqual(report.per_lot_evaluation[lot][p]['144h'], undefined);
      }
    }
    console.log('  -> PASS: Attack G');
  }

  // Attack H: Missing required parameter rejected
  console.log('Running Attack H: Missing required parameter rejected fail-closed...');
  {
    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'predicta-attack-h-'));
    const contract = loadAuthoritativeStabilityContract();
    const contractMod = JSON.parse(JSON.stringify(contract));
    contractMod.methodology.target_parameters = ['ileak', 'tpd']; // iddq missing!

    const modPath = path.join(tmpDir, 'mod_contract.json');
    fs.writeFileSync(modPath, JSON.stringify(contractMod));

    assert.throws(
      () => new MultiLotConformalStabilityEvaluator({ stabilityContractPath: modPath }),
      /MISSING_REQUIRED_PARAMETER/
    );
    fs.rmSync(tmpDir, { recursive: true, force: true });
    console.log('  -> PASS: Attack H');
  }

  // Attack I: Wrong nominal level rejected
  console.log('Running Attack I: Wrong nominal level rejected...');
  {
    const contract = loadAuthoritativeStabilityContract();
    assert.deepStrictEqual(contract.methodology.candidate_nominal_levels, [0.8, 0.9, 0.95]);
    console.log('  -> PASS: Attack I');
  }

  // Attack J: Deterministic repeated evaluation
  console.log('Running Attack J: Deterministic repeated evaluation...');
  {
    const rep1 = evaluator.evaluate();
    const rep2 = evaluator.evaluate();

    for (const p of ['iddq', 'ileak', 'tpd']) {
      for (const h of ['96h', '168h']) {
        for (const lvl of ['0.80', '0.90', '0.95']) {
          const a1 = rep1.aggregate_evaluation[p][h][lvl];
          const a2 = rep2.aggregate_evaluation[p][h][lvl];
          assert.strictEqual(a1.empirical_coverage_pct, a2.empirical_coverage_pct);
          assert.strictEqual(a1.covered_count, a2.covered_count);
          assert.strictEqual(a1.conformal_quantile_q, a2.conformal_quantile_q);

          const d1 = rep1.cross_lot_dispersion[p][h][lvl];
          const d2 = rep2.cross_lot_dispersion[p][h][lvl];
          assert.strictEqual(d1.min_lot_coverage_pct, d2.min_lot_coverage_pct);
          assert.strictEqual(d1.max_lot_coverage_pct, d2.max_lot_coverage_pct);
          assert.strictEqual(d1.lot_coverage_range_pct, d2.lot_coverage_range_pct);
          assert.strictEqual(d1.std_lot_coverage_pct, d2.std_lot_coverage_pct);
        }
      }
    }
    console.log('  -> PASS: Attack J');
  }

  // Attack K: Provenance mismatch rejected
  console.log('Running Attack K: Provenance mismatch / artifact tampering rejected...');
  {
    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'predicta-attack-k-'));
    const art = JSON.parse(fs.readFileSync(CALIBRATION_ARTIFACT_PATH, 'utf8'));
    art.conformal_quantiles.iddq['96h']['0.80'] = 9999.0; // tamper!
    const tamperedArt = path.join(tmpDir, 'tampered.json');
    fs.writeFileSync(tamperedArt, JSON.stringify(art));

    assert.throws(
      () => new MultiLotConformalStabilityEvaluator({ calibrationArtifactPath: tamperedArt }),
      /CALIBRATION_ARTIFACT_TAMPERING_DETECTED/
    );
    fs.rmSync(tmpDir, { recursive: true, force: true });
    console.log('  -> PASS: Attack K');
  }

  // Attack L: Aggregate coverage cannot hide lot records
  console.log('Running Attack L: Aggregate coverage cannot hide lot records...');
  {
    const report = evaluator.evaluate();
    assert.strictEqual(Object.keys(report.per_lot_evaluation).length, 8);
    for (const p of ['iddq', 'ileak', 'tpd']) {
      for (const h of ['96h', '168h']) {
        const d = report.cross_lot_dispersion[p][h]['0.80'];
        assert.ok(d.lot_coverage_range_pct > 0.0, 'Range across lots must be greater than 0');
        assert.ok(d.min_lot_coverage_pct < d.max_lot_coverage_pct);
      }
    }
    console.log('  -> PASS: Attack L');
  }

  // Attack M: No arbitrary production acceptance threshold
  console.log('Running Attack M: No arbitrary production acceptance threshold...');
  {
    const report = evaluator.evaluate();
    assert.strictEqual(report.governance_status.governance_status, 'REVIEW_REQUIRED');
    assert.strictEqual(
      report.governance_status.acceptance_threshold_status,
      'NO_PRODUCTION_ACCEPTANCE_THRESHOLD_AUTHORIZED'
    );
    assert.strictEqual(report.governance_status.promotion_locked, true);
    console.log('  -> PASS: Attack M');
  }

  // Attack N: BENCHMARK_ONLY status unchanged
  console.log('Running Attack N: BENCHMARK_ONLY status unchanged...');
  {
    const report = evaluator.evaluate();
    assert.strictEqual(report.governance_status.model_status, 'BENCHMARK_ONLY');
    console.log('  -> PASS: Attack N');
  }

  // Attack O: NOT_CALIBRATED status unchanged
  console.log('Running Attack O: NOT_CALIBRATED status unchanged...');
  {
    const report = evaluator.evaluate();
    assert.strictEqual(report.governance_status.calibration_status, 'NOT_CALIBRATED');
    console.log('  -> PASS: Attack O');
  }

  console.log('='.repeat(80));
  console.log('ALL 15 STAGE 6 TASK 3 ATTACKS (A-O) PASSED CLEANLY (NODE.JS)');
  console.log('='.repeat(80));
}

if (require.main === module) {
  runTests();
}

module.exports = { runTests };
